factcast-factus/src/test/java/org/factcast/factus/applier/DefaultProjectorTest.java

```diff
@@ -0,0 +1,519 @@
+            List<FactSpec> factSpecs = underTest.createFactSpecs();
+
+            // ASSERT
+            assertThat(factSpecs)
+                    .hasSize(2)
+                    .flatExtracting(
+                            FactSpec::aggId,
+                            FactSpec::ns,
+                            FactSpec::version,
+                            FactSpec::type)
+                    .contains(
+                            // ComplexAggregate has two handlers
+                            aggregateId, "test", 0, "ComplexEvent",
+                            aggregateId, "test", 0, "ComplexEvent2");
+        }
+
+        @Test
+        void createFromProjectionWithHandlerFor() {
+            // INIT
+            ProjectionWithHandlerFor projection = new ProjectionWithHandlerFor();
+
+            DefaultProjector<ProjectionWithHandlerFor> underTest = new DefaultProjector<>(
+                    eventSerializer,
+                    projection);
+
+            // RUN
+            List<FactSpec> factSpecs = underTest.createFactSpecs();
+
+            // ASSERT
+            assertThat(factSpecs)
+                    .hasSize(1)
+                    .flatExtracting(
+                            FactSpec::aggId,
+                            FactSpec::ns,
+                            FactSpec::version,
+                            FactSpec::type)
+                    .contains(
+                            null, "test", 0, "someType");
+        }
+
+        @Test
+        void invalidPostprocessReturnsNull() {
+            // INIT
+            DefaultProjector<Projection> underTest = new DefaultProjector<>(
+                    eventSerializer,
+                    new PostProcessingProjection(null));
+
+            // RUN
+            assertThatThrownBy(() -> underTest.createFactSpecs())
+                    // ASSERT
+                    .isInstanceOf(IllegalArgumentException.class)
+                    .hasMessageStartingWith("No FactSpecs discovered from");
+        }
+
+        @Test
+        void invalidPostprocessReturnsEmptyList() {
+            // INIT
+            DefaultProjector<Projection> underTest = new DefaultProjector<>(
+                    eventSerializer,
+                    new PostProcessingProjection(Lists.emptyList()));
+
+            // RUN
+            assertThatThrownBy(() -> underTest.createFactSpecs())
+                    // ASSERT
+                    .isInstanceOf(IllegalArgumentException.class)
+                    .hasMessageStartingWith("No FactSpecs discovered from");
+        }
+
+    }
+
+    @Nested
+    class whenConstructingWithDefectiveProjection {
+        @Test
+        void duplicateHandler() {
+            // INIT / RUN
+            assertThatThrownBy(() -> new DefaultProjector<>(
+                    eventSerializer,
+                    new DuplicateHandlerProjection()))
+                            // ASSERT
+                            .isInstanceOf(UnsupportedOperationException.class)
```

Comment:
pondering if exception type should be changed in this situation... suggestions?

Reply:
Something fancy like UnsatisfiedLinkError or LayerInstantiationException? :P 

I agree, I suggest to create your own exception type though